cmd/kubeadm/app/master/selfhosted_volumes.go

```diff
@@ -0,0 +1,294 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package master
+
+import (
+	"fmt"
+	"io/ioutil"
+	"path"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+)
+
+func apiServerProjectedVolume() *v1.Volume {
+	return &v1.Volume{
+		Name: "k8s",
+		VolumeSource: v1.VolumeSource{
+			Projected: &v1.ProjectedVolumeSource{
+				Sources: []v1.VolumeProjection{
+					{
+						Secret: &v1.SecretProjection{
+							LocalObjectReference: v1.LocalObjectReference{
+								Name: kubeadmconstants.CACertAndKeyBaseName,
+							},
+							Items: []v1.KeyToPath{
+								{
+									Key:  "tls.crt",
+									Path: path.Join("pki", kubeadmconstants.CACertName),
+								},
+								{
+									Key:  "tls.key",
+									Path: path.Join("pki", kubeadmconstants.CAKeyName),
+								},
+							},
+						},
+					},
+					{
+						Secret: &v1.SecretProjection{
+							LocalObjectReference: v1.LocalObjectReference{
+								Name: kubeadmconstants.APIServerCertAndKeyBaseName,
+							},
+							Items: []v1.KeyToPath{
+								{
+									Key:  "tls.crt",
+									Path: path.Join("pki", kubeadmconstants.APIServerCertName),
```

Comment:
This locks us to a certificate directory that ends with `pki`. We want the cert directory to be set by the user to anything.
I'd say that it's better to call this volume `certificates` or `certs` or whatever and make it not conflict with the "normal" `/etc/kubernetes` mount.

Here I'd prefer just `kubeadmconstants.APIServerCertName`